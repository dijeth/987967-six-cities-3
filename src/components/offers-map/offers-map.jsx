import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import leaflet from 'leaflet';
import {isEqualCoords} from '../../util.js';
import {connect} from 'react-redux';
import {getActiveOfferCoord} from '../../reducers/app/selectors.js';

const ICON_SIZE = [27, 39];

const ICON = leaflet.icon({
  iconUrl: `img/pin.svg`,
  iconSize: ICON_SIZE
});

const ICON_ACTIVE = leaflet.icon({
  iconUrl: `img/pin-active.svg`,
  iconSize: ICON_SIZE
});

class OffersMap extends PureComponent {
  constructor(props) {
    super(props);

    this._centerCoord = null;
    this._activeCoord = null;
    this._offersCoord = [];
    this._activeLayer = null;
    this._offerLayers = [];

    this._mapWrapper = React.createRef();
  }

  componentDidMount() {
    const {centerCoord, offersCoord, activeCoord, zoom} = this.props;

    this._mapWrapper.current.style.height = `100%`;

    this._map = leaflet.map(this._mapWrapper.current, {
      center: centerCoord,
      zoom,
      zoomControl: false,
      marker: true
    });

    leaflet
      .tileLayer(
          `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
            attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
          })
      .addTo(this._map);

    this._setView(centerCoord, zoom);

    this._centerCoord = centerCoord;
    this._updateOffers(offersCoord);
    this._updateActiveCoord(activeCoord);
  }

  componentDidUpdate() {
    const {centerCoord, offersCoord, activeCoord, zoom} = this.props;

    if (this._shouldOfferMarkersBeChanged(centerCoord, offersCoord)) {
      this._centerCoord = centerCoord;

      this._setView(centerCoord, zoom);
      this._updateOffers(offersCoord);
    }

    this._updateActiveCoord(activeCoord);
  }

  componentWillUnmount() {
    this._map.remove();
  }

  _shouldOfferMarkersBeChanged(centerCoord, offersCoord) {
    if (!isEqualCoords(centerCoord, this._centerCoord)) {
      return true;
    }

    if (offersCoord.length !== this._offersCoord.length) {
      return true;
    }

    if (offersCoord.filter((it, i) => !isEqualCoords(it, this._offersCoord[i])).length) {
      return true;
    }

    return false;
  }

  _setView(coords, zoom) {
    this._map.setView(coords, zoom);
  }

  _addOffers(offersCoord) {
    this._offersCoord = offersCoord;
    return offersCoord.map((coord) => leaflet.marker(coord, {icon: ICON}).addTo(this._map));
  }

  _updateOffers(offersCoord) {
    this._offerLayers.forEach((it) => {
      this._map.removeLayer(it);
    });

    this._offerLayers = null;
    this._offerLayers = this._addOffers(offersCoord);
  }

  _updateActiveCoord(coord) {
    if (this._activeLayer !== null) {
      this._map.removeLayer(this._activeLayer);
      this._activeLayer = null;
    }

    if (coord) {
      this._activeLayer = leaflet.marker(coord, {icon: ICON_ACTIVE}).addTo(this._map);
    }
  }

  render() {
    return <div id='map' ref={this._mapWrapper}></div>;
  }
}

const mapStateToProps = (state) => ({
  activeCoord: getActiveOfferCoord(state)
});

OffersMap.propTypes = {
  centerCoord: PropTypes.arrayOf(PropTypes.number).isRequired,
  offersCoord: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,
  activeCoord: PropTypes.arrayOf(PropTypes.number),
  zoom: PropTypes.number
};

export {OffersMap};
export default connect(mapStateToProps)(OffersMap);
