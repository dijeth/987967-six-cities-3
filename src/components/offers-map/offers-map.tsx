import * as React from 'react';
import * as leaflet from 'leaflet';
import {isEqualCoords} from '../../util.js';
import {connect} from 'react-redux';
import {getActiveOfferCoord} from '../../reducers/app/selectors.js';
import {coord} from '../../types';

const ICON_SIZE: [number, number] = [27, 39];

const ICON = leaflet.icon({
  iconUrl: `img/pin.svg`,
  iconSize: ICON_SIZE
});

const ICON_ACTIVE = leaflet.icon({
  iconUrl: `img/pin-active.svg`,
  iconSize: ICON_SIZE
});

const LAYER_URL = `https://{s}.basemaps.cartocdn.com/rastertiles/voyager_labels_under/{z}/{x}/{y}{@2x}.png`;

type Props = {
  centerCoord: coord;
  offersCoord: Array<coord>;
  activeCoord: coord;
  zoom: number;
};

class OffersMap extends React.PureComponent<Props> {
  private mapRef: React.RefObject<HTMLDivElement>;
  private map: leaflet.Map;
  private layerGroup: leaflet.LayerGroup;
  private activeLayer: leaflet.Marker;

  private offersCoord: Array<coord>;
  private activeCoord: coord;
  private centerCoord: coord;

  constructor(props) {
    super(props);
    this.mapRef = React.createRef();
    this.map = null;

    this.layerGroup = null;
    this.activeLayer = null;

    this.offersCoord = null;
    this.activeCoord = null;
    this.centerCoord = null;
  }

  updateActiveLayer() {
    if (
      JSON.stringify(this.activeCoord) ===
      JSON.stringify(this.props.activeCoord)
    ) {
      return;
    }

    this.activeCoord = this.props.activeCoord;

    if (this.activeLayer !== null) {
      this.activeLayer.remove();
      this.activeLayer = null;
    }

    if (this.activeCoord !== null) {
      this.activeLayer = leaflet
        .marker(this.activeCoord, { icon: ICON_ACTIVE })
        .addTo(this.map);
    }
  }

  updateLayerGroup() {
    if (
      JSON.stringify(this.offersCoord) ===
      JSON.stringify(this.props.offersCoord)
    ) {
      return;
    }

    this.offersCoord = this.props.offersCoord;
    this.layerGroup.clearLayers();

    this.offersCoord.forEach(it => {
      leaflet.marker(it, { icon: ICON }).addTo(this.layerGroup);
    });
  }

  updateView() {
    if (
      JSON.stringify(this.centerCoord) ===
      JSON.stringify(this.props.centerCoord)
    ) {
      return;
    };

    this.centerCoord = this.props.centerCoord;
    this.map.setView(this.centerCoord, this.props.zoom);
  }

  componentDidUpdate() {
    this.updateActiveLayer();
    this.updateLayerGroup();
    this.updateView();
  }

  componentDidMount() {
    this.map = leaflet.map(this.mapRef.current.id, {
      zoomControl: false,
      scrollWheelZoom: false
    });

    leaflet.tileLayer(LAYER_URL).addTo(this.map);
    
    this.layerGroup = leaflet.layerGroup().addTo(this.map);

    this.updateActiveLayer();
    this.updateLayerGroup();
    this.updateView();
  }

  render() {
    return (
      <div
        ref={this.mapRef}
        id="map"
        style={{ width: `100%`, height: `100%` }}
      ></div>
    );
  }
}

const mapStateToProps = (state) => ({
  activeCoord: getActiveOfferCoord(state)
});

export {OffersMap};
export default connect(mapStateToProps)(OffersMap);
