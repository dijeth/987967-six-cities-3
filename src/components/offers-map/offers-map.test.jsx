import React from 'react';
import Enzyme, {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import {OffersMap} from './offers-map.jsx';

Enzyme.configure({
  adapter: new Adapter()
});

const props = {
  centerCoord: [52.38333, 4.9],
  offersCoord: [
    [52.359385, 4.879898],
    [52.353995, 4.911789],
    [52.377303, 4.903075]
  ],
  activeCoord: [52.377304, 4.903076]
};

it(`<OffersMap /> should be render correctly with activeCoord`, () => {
  const div = global.document.createElement(`div`);
  global.document.body.appendChild(div);

  const map = mount(<OffersMap {...props} />, {attachTo: div});
  expect(toJson(map, {mode: `shallow`})).toMatchSnapshot();
});


