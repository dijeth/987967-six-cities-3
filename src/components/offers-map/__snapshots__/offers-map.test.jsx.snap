// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OffersMap /> should be render correctly with activeCoord 1`] = `
"<OffersMap centerCoord={{...}} offersCoord={{...}} activeCoord={{...}}>
  <div id=\\"map\\" />
</OffersMap>"
`;
